contrib/epee/include/serialization/wire/read.h

```diff
@@ -0,0 +1,532 @@
+    };
+
+    //! \return Maximum read depth for both objects and arrays before erroring
+    static constexpr std::size_t max_read_depth() noexcept { return 100; }
+
+    //! \return Assume delimited arrays in generic interface (some optimizations disabled)
+    static constexpr std::true_type delimited_arrays() noexcept { return {}; }
+
+    virtual ~reader() noexcept
+    {}
+
+    //! \return Number of recursive objects and arrays
+    std::size_t depth() const noexcept { return depth_; }
+
+    //! \return Unprocessed bytes
+    epee::span<const std::uint8_t> remaining() const noexcept { return remaining_; }
+
+    //! \throw wire::exception if parsing is incomplete.
+    virtual void check_complete() const = 0;
+
+    //! \throw wire::exception if array, object, or end of stream.
+    virtual basic_value basic() = 0;
+
+    //! \throw wire::exception if next value not a boolean.
+    virtual bool boolean() = 0;
+
+    //! \throw wire::expception if next value not an integer.
+    virtual std::intmax_t integer() = 0;
+
+    //! \throw wire::exception if next value not an unsigned integer.
+    virtual std::uintmax_t unsigned_integer() = 0;
+
+    //! \throw wire::exception if next value not number
+    virtual double real() = 0;
+
+    //! throw wire::exception if next value not string
+    virtual std::string string() = 0;
+
+    /*! Copy upcoming string directly into `dest`.
+      \throw wire::exception if next value not string
+      \throw wire::exception if next string exceeds `dest.size())`
+      \throw wire::exception if `exact == true` and next string is not `dest.size()`
+      \return Number of bytes read into `dest`. */
+    virtual std::size_t string(epee::span<char> dest, bool exact) = 0;
+
+    // ! \throw wire::exception if next value cannot be read as binary
+    virtual epee::byte_slice binary() = 0;
+
+    /*! Copy upcoming binary directly into `dest`.
+      \throw wire::exception if next value not binary
+      \throw wire::exception if next binary exceeds `dest.size()`
+      \throw wire::exception if `exact == true` and next binary is not `dest.size()`.
+      \return Number of bytes read into `dest`. */
+    virtual std::size_t binary(epee::span<std::uint8_t> dest, const bool exact) = 0;
+
+    /*! \param min_element_size of each array element in any format - if known.
+          Derived types with explicit element count should verify available
+          space, and throw a `wire::exception` on issues.
+        \throw wire::exception if next value not array
+        \throw wire::exception if not enough bytes for all array elements
+          (with epee/msgpack which has specified number of elements).
+        \return Number of values to read before calling `is_array_end()`. */
+    virtual std::size_t start_array(std::size_t min_element_size) = 0;
+
+    //! \return True if there is another element to read.
+    virtual bool is_array_end(std::size_t count) = 0;
+
+    void end_array() noexcept { decrement_depth(); }
```

Comment:
I would mark this as `virtual` in case it is ever needed to be overridden by a subclass. Since we usually use `reader` in a templated context, the compiler should usually be able to de-virtualize this method most of the time anyways. Same with `end_object`.

Reply:
I haven't had a need to `override` for msgpack, epee, protobuf, or JSON (and unlikely for CBOR either). `is_array_end(size_t)` and `key(...)` are both `virtual`, and are used to determine the end whereas these functions are simply used by the base class to track placement in the tree.

Reply:
Also, this is trivial to change this if a future format needs to override it.